dts: sensor support and i2c bus generation

erwango opened the pull request:
This pull-request provides support for sensor in device tree.

New yaml file is provided for sensor.
To help generation of a I2C bus master information, a new type of yaml property category is defined
with new yaml category: default-value. More information on this provided in commit messages.

The pull request focus on disco_l475_iot1 board as feature enabler

erwango committed:
dts: bindings: provide generic yaml description for i2c-devices

Provide generic yaml description for i2c-devices such as sensors

New yaml binding category 'parent/child' is added.
It aims at binding two types of related node such as bus master and
bus slave.
In case of i2c-device object, parent property is 'bus' with value 'i2c'.
In the mean time, i2c node gets child property bus. Master and slave bus
values should match.
As such, 'i2c' node is bus master and 'i2c-devices' nodes are bus slaves

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

galak commented:
DId you see other use cases for 'required-implicit'?

galak commented on drivers/sensor/lis3mdl/Kconfig:
```diff
@@ -39,10 +39,9 @@ config LIS3MDL_I2C_MASTER_DEV_NAME
+choice LIS3MDL_TRIGGER_MODE
 	prompt "Trigger mode"
 	depends on LIS3MDL
-	default LIS3MDL_TRIGGER_GLOBAL_THREAD
```
any reason to remove the default?

erwango replied:
Yes, I think I explained this in the commit. This is required (at least I couldn't do otherwise) to be able to define a value in board Kconfig.defconfig

MaureenHelm replied:
The commit message explains why you added a name to the choice, not why you removed the default.

erwango replied:
This is right! 
So actually this is the combination of 2 (remove default and provide name to the choice) which was required.
Unfortunately, this commit has already been merged in arm branch, so I won't be able to make it clearer.
(And current commit should be remove when rebasing)

galak commented on arch/arm/soc/st_stm32/stm32f1/Kconfig.series:
```diff
@@ -14,5 +14,6 @@ config SOC_SERIES_STM32F1X
 	select HAS_STM32CUBE
 	select CPU_HAS_SYSTICK
 	select CLOCK_CONTROL_STM32_CUBE if CLOCK_CONTROL
+	select I2C_STM32_V1 if I2C
```
I think i moved this out of here take a look at:

commit 8e8d49a6202a472513354b1db27d2b4ffd23faa5
Author: Kumar Gala <kumar.gala@linaro.org>
Date:   Wed Jul 5 11:04:06 2017 -0500

    i2c: stm32: Cleanup how we enable the specific I2C driver

galak commented on dts/common/yaml/sensor.yaml:
```diff
@@ -0,0 +1,30 @@
+---
+title: Sensor Base Structure
```
this could probably be more generically just an i2c device.  Nothing sensor specific

erwango replied:
so i2c-device.yaml?

galak replied:
yes

galak commented on dts/common/yaml/sensor.yaml:
```diff
@@ -0,0 +1,30 @@
+      description: compatible strings
+  - reg:
+      type: array
+      description: mmio register space
```
description is wrong, should be i2c bus address

erwango replied:
ok

galak commented on dts/common/yaml/sensor.yaml:
```diff
@@ -0,0 +1,30 @@
+      description: mmio register space
+      generation: define
+      category: required
+  - label:
```
can we make label optional

agross-linaro replied:
So in Linux, SPI and I2C have implicit busses where the bus is indexed either first to last or via the alias.  So correct me if I am wrong, but the whole intent of this is to fixate the bus names so that the slave devices can be bound properly and applications can get at the right device?

erwango replied:
"indexed either first to last"
I'm not sure to get the idea, but let's say it's defined by the parent node.
"So correct me if I am wrong, but the whole intent of this is to fixate the bus names so that the slave devices can be bound properly and applications can get at the right device?"
yes

galak commented on dts/common/yaml/device_node.yaml.template:
```diff
@@ -13,10 +13,11 @@ properties:
 #     type: <string | int | array>
 #     description: <description of property>
 #     generation: <define | structure>
+#     default-value: keyword for required-implicit generation
```
what values do we support for 'default-value'?

erwango replied:
Foe now: parent-label

galak reviewed (changes requested):
Some feedback

galak commented:
> Not yet. I wanted t avoid a too much ad-hoc, and this was the best solution I've found.
> I wanted to add it as a yaml property as this is a node property that what we want to generate and I found that using other solutions such as #cells or something not under "properties" would end up being not generic enough.
> 'required' would have been confusing, as property is not there in general.
> 'implicit' would not be true either as i2c-bus actually exists and could be used for this purpose.
> 'required': was not quite

I wonder about doing just "required-parent”, and dropping the "default-value” bit.

- k

erwango commented:
> I wonder about doing just "required-parent”, and dropping the "default-value” bit.`

What about using "optional" + default-value?
With just "required-parent" I feel this would end up in a ad-hoc code where required-parent would bound to "get parent label". Besides, there is a possibility to actually state i2c-bus (according to Linux documentation). In that case you need to use provided value.

erwango commented:
@galak: one point I was not sure, is location of hts221.yaml and others.
Under /arm/ right now, I feel it is not appropriate

galak commented:
> @galak: one point I was not sure, is location of hts221.yaml and others.
> Under /arm/ right now, I feel it is not appropriate

Agreed, I want to restructure the yaml dirs to try and mimic the linux dt binding structure as much as possible.  So if those sensors have reasonable locations from a linux dt binding point of view, that we should put them there.

erwango commented:
> Agreed, I want to restructure the yaml dirs to try and mimic the linux dt binding structure as much as possible. So if those sensors have reasonable locations from a linux dt binding point of view, that we should put them there.

Sounds good to me

dbkinder commented on drivers/sensor/hts221/Kconfig:
```diff
@@ -16,23 +16,22 @@ config HTS221_NAME
-	depends on HTS221
+	depends on HTS221 && !HAS_DTS_SENSOR
 	help
 	  Device name with which the HTS221 sensor is identified.
```
I'd like to update these help texts, but they're not part of your patch. Should I make a note and come back later to fix them?  For example, this would read better as:

     Device name identifying the HTS221 sensor.

erwango replied:
I had a look and it seems basically all sensors require an update.
If possible I'd like to avoid doing it inside this PR :-)

dbkinder commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -196,6 +196,18 @@ def find_parent_irq_node(node_address):
+    if prop in reduced[address]['props']:
+        parent_prop = reduced[address]['props'].get(prop)
+    else:
+        raise("Parent doesn't have requested property")
```
will this error provide enough context information to know what to fix?  Which node?  What property?

erwango replied:
Right, could be enhanced easily.

erwango commented:
@galak: can you feedback on this new version?

galak commented:
I pulled in the base set of patches that enabled the sensors on the disco board.

galak commented:
So from a yaml point of view, I'm wanting to maybe add a new property to 'reg' to qualify it as on a i2c-bus.  Wondering if/how that might work/look.

from the 'yaml' side we'd only have something like (with 'bus') being the new thing.

+  - reg:
 +      type: array
 +      description: address on i2c bus
 +      generation: define
 +      category: required
 +      bus: i2c

erwango commented:
Like :
if 'reg' property has 'bus' attribute: 
   * add a 'bus' property to node
   * 'bus' value = parent node label

This should be doable(I guess we can do anything), but might be a little bit trickier since we need to generate 2 defs base on reg attributes. This require extra logic in extract_reg_prop which is not yaml aware right now  (today, it works only with #address-cells and #size-cells).

I feel this would be a bit awkward, my view is we want to generate a #define element out of the node.
This is done generally thanks to treatment of node properties described as yaml 'properties' of a yaml node binding. Exception is when we generate more than one #define out of a single property, which is done thanks to "#cells" hook.
It would be nice if we don't add a new method each time we have a new need. This what I tried to do by trying to re-use existing node property generic treatment. I'm sure it could be done otherwise, and more generic hopefully. But I'd like to avoid doing it in a less generic way.

galak commented:
> Like :
> if 'reg' property has 'bus' attribute:
> 
> add a 'bus' property to node
> 'bus' value = parent node label
> This should be doable(I guess we can do anything), but might be a little bit trickier since we need to generate 2 defs base on reg attributes. This require extra logic in extract_reg_prop which is not yaml aware right now (today, it works only with #address-cells and #size-cells).
> 
> I feel this would be a bit awkward, my view is we want to generate a #define element out of the node.
> This is done generally thanks to treatment of node properties described as yaml 'properties' of a yaml node binding. Exception is when we generate more than one #define out of a single property, which is done thanks to "#cells" hook.
> It would be nice if we don't add a new method each time we have a new need. This what I tried to do by trying to re-use existing node property generic treatment. I'm sure it could be done otherwise, and more generic hopefully. But I'd like to avoid doing it in a less generic way.

So I agree we don't want to add a new method every time, but I think we have some patterns we will find out and need to add some hooks/changes.  I was looking at this and think we can do something like in extract_property():

if y_key == 'reg':
            extract_reg_prop(node_address, names, defs, def_label,
                          1, y_val.get('label', None))
            bus = y_val.get('bus', None)
            if bus:
                 ....

galak commented:
I pushed an update to the branch with a change that supports 'bus' as I described.  We probably want to clean this up, but its at least in a working state.

galak commented:
Another issue we need to deal with is have the same sensor on 2 different buses.  This is broken right now based on how we generate the #define names.

erwango commented:
I'll have a look to the change.
By the way, what about existing i2c-bus property, as (sparely) used in Linux.

> Another issue we need to deal with is have the same sensor on 2 different buses.

Do you have an example on how this happens in Linux dts?

galak commented:
Here's a snipet based on disco:


```
&i2c1 {
        status = "ok";
        clock-frequency = <I2C_BITRATE_FAST>;

        lis3mdl-magn@1e {
                compatible = "st,lis3mdl-magn";
                reg = <0x1e>;
                label = "LIS3MDL";
        };

};

&i2c2 {
        status = "ok";
        clock-frequency = <I2C_BITRATE_FAST>;

        lis3mdl-magn@1e {
                compatible = "st,lis3mdl-magn";
                reg = <0x1e>;
                label = "LIS3MDL";
        };


```

erwango commented:
ok, yes, was elementary.

erwango commented:
> Another issue we need to deal with is have the same sensor on 2 different buses

I understand these would be 2 different sensors. Each one being located on a different bus.
In that case, we need to add bus id/base address in sensor define.

erwango commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -494,6 +494,12 @@ def extract_property(node_compat, yaml, node_address, y_key, y_val, names,
+        bus = y_val.get('bus', None)
+        if bus:
+            extract_single(node_address, yaml,
+                           'parent-label', bus + '-bus',
```
'parent-label' sounds like a hook in the hook here.
Should be cleaned up somehow.

erwango commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -494,6 +494,12 @@ def extract_property(node_compat, yaml, node_address, y_key, y_val, names,
                          1, y_val.get('label', None))
+        bus = y_val.get('bus', None)
+        if bus:
+            extract_single(node_address, yaml,
```
This is essentially why I dislike this solution.
We end up with lot of ad hoc triage in this function with use of specific extract_<new property>, while I hope we should be able to reduce it thanks to generic scheme that should be provided by yaml description.

erwango commented:
If we're to add bus id/bus base address in i2c-device node #defines, first idea would be: 
#define ST_STM32_I2C_V2_40005400_ST_HTS221_5F_BASE_ADDRESS0	0x5f
#define ST_STM32_I2C_V2_40005400_ST_HTS221_5F_I2C_BUS		"I2C_2"
#define ST_STM32_I2C_V2_40005400_ST_HTS221_5F_LABEL		"HTS221"
#define ST_STM32_I2C_V2_40005400_ST_HTS221_5F_BASE_ADDRESS	ST_STM32_I2C_V2_40005400_ST_HTS221_5F_BASE_ADDRESS_0

This requires new mecanism. Maybe alike of current "cell_string"/"#cells", that would be added in i2c.yaml, which would then need to be included in i2c-device.yaml. For instance (to be refined):

bus: I2C_BUS
"#cells":
  - master_label

galak commented:
> If we're to add bus id/bus base address in i2c-device node #defines, first idea would be:
> #define ST_STM32_I2C_V2_40005400_ST_HTS221_5F_BASE_ADDRESS0	0x5f
> #define ST_STM32_I2C_V2_40005400_ST_HTS221_5F_I2C_BUS	"I2C_2"
> #define ST_STM32_I2C_V2_40005400_ST_HTS221_5F_LABEL	"HTS221"
> #define ST_STM32_I2C_V2_40005400_ST_HTS221_5F_BASE_ADDRESS ST_STM32_I2C_V2_40005400_ST_HTS221_5F_BASE_ADDRESS_0
> 
> This requires new mecanism. Maybe alike of current "cell_string"/"#cells", that would be added in i2c.yaml, which would then need to be included in i2c-device.yaml. For instance (to be refined):
> 
> bus: I2C_BUS
> "#cells":
> 
> master_label

I was thinking something like this, instead of adding 'bus' to reg property in the yaml, of introducing something new that conveys something about the parent.  So maybe:

description:

parent:
    - bus: i2c

properties:

erwango commented:
Pushed functionnal attempt for parent-node yaml section.
Still WIP, as master bus generation code is quite specific.

MaureenHelm commented on dts/bindings/i2c/i2c-device.yaml:
```diff
@@ -0,0 +1,30 @@
+      generation: define
+  - i2c-bus:
+      type: string
+      category: optionnal
```
s/optionnal/optional/

MaureenHelm commented on dts/bindings/i2c/i2c-device.yaml:
```diff
@@ -0,0 +1,30 @@
+---
```
Add license header

erwango replied:
ok, it seems it's missing in bunch of them.
Will do something

erwango replied:
@MaureenHelm: What copyright should be provided for files that I didn't created?
Should I look for the original author and put his copyright?
Or can I just put SPDX boiler plate w/o any copyright?

MaureenHelm replied:
Use your own copyright for the new files you add

MaureenHelm commented on dts/Kconfig:
```diff
@@ -19,3 +19,10 @@ config HAS_DTS_I2C
 	This option specifies that the target platform supports device tree
 	configuration for I2c.
 
+config HAS_DTS_SENSOR
```
I think this should now be `HAS_DTS_I2C_DEVICE`?

erwango replied:
Good catch!

erwango commented:
@galak: Before I do some clean up, how do you feel with this new proposal?

On my side, I'm not fully satisfied with the way we retrieve label information from parent to child (both on yaml side and python side). Any idea if there would be some other king of info we'd need to get from parent ?
If this is only label, maybe we could just assume that if node parent is bus, generate #define I2C_BUS_MASTER.. by default.

galak commented on dts/bindings/i2c/i2c-device.yaml:
```diff
@@ -0,0 +1,33 @@
+
+parent-node:
+  - bus: i2c
+  - master: label
```
what's 'master: label' suppose to mean?

galak commented on dts/bindings/device_node.yaml.template:
```diff
@@ -17,6 +17,8 @@ properties:
 #     type: <string | int | array>
 #     description: <description of property>
 #     generation: <define | structure>
+#     default-value: used with "category: optional" and when property is not
```
do we need this anymore, doesn't seem like we are using it.

galak commented on dts/bindings/i2c/i2c-device.yaml:
```diff
@@ -0,0 +1,33 @@
+description: >
+    This binding gives the base structures for all i2c devices
+
+parent-node:
```
I think we can just say 'parent' w/o the -node

erwango commented on dts/bindings/i2c/i2c-device.yaml:
```diff
@@ -0,0 +1,33 @@
+    This binding gives the base structures for all i2c devices
+
+parent-node:
+  - bus: i2c
```
I think here we can do a bit more simple here:
parent-node: bus-master

Then in the script, test this single condition:
if parent-node is bus-master:
  *generate defs with <PARENT_BASE_LABEL>_ prefix
  *generate def <PARENT_BASE_LABEL>_I2C_BUS = parent label

erwango committed:
dts: bindings: add bindings for 4 sensors

Provide dts yaml bindings for 4 sensors supported by disco_l475_iot1
board:
-hts221, lis3mdl, lps22hb and lsm6dsl
Devices are denoted as i2c-devices and, as such, inherits from
i2c-device.yaml

Signed-Off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
scripts: extract_dts_include: update to match new bindings path

Following move of dts bindings from common/yaml to bindings,
update yaml file search routine.

Signed-Off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts: add sensors dts support for some sensors

Provide HAS_DTS_I2C_DEVICE flag to enable dts based generation for
these sensors.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
scripts: extract_dts_inculdes: treat 'parent: bus' attribute

When node has 'bus' as 'parent' attribute, change base label to
include parent address in label prefix.

Besides, generates a "_BUS_NAME" define which equals parent label.

Signed-Off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
board: disco_l475_iot1: Use dts for sensor definition

Move sensor related configuration to dts and dts based
generated defines on disco_l475_iot1.

Signed-Off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
@galak, @MaureenHelm, @agross-linaro : PR Cleaned-up.
Branch is based onto #5076.

SebastianBoe commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -508,16 +551,15 @@ def extract_node_include_info(reduced, root_node_address, sub_node_address,
+            label = y_node['properties']['label']
+            label_override = convert_string_to_label(
+                                    node['props']['label']).upper()
+        except:
```
except should always be filtered with an exception type so you don't get an unexpected exception (e.g. OOM, syntax error) and have a hard time debugging.

erwango replied:
Actually, I as trying to avoid 2 steps check to see if y_node has ['properties']['label'].
If it hasn't, it's not actually an error, we just need to go on with last part of the function.
Would there be another way to do that or am I stuck with 2 steps check?

SebastianBoe replied:
Then I would do "except KeyError" at least.

https://stackoverflow.com/a/15036813/1134134

erwango replied:
ok, Thanks!
Then maybe the whole script would need a clean up then.
I'll do it in a separate commit

erwango commented:
recheck

erwango commented:
Moved to master branch as arm branch is rebased on master branch by CI before testing and it showed some conflicts at that step

galak commented:
I'd like a means for us to associate the parent and child yaml's together for things on a bus.  So I'm thinking something like:

[ in dts/bindings/i2c/i2c-device.yaml ]

parent:
   bus: i2c
      generation: define

[ in dts/bindings/i2c/i2c.yaml ]

child:
  bus: i2c
  
If 'parent' is set for a node's yaml, that we should check that the parent node in the yaml has 'bus' set.  If that's true, than the values of 'bus' should be the same.  If either case isn't true, its a parsing error.

erwango commented:
@galak : So added a new commit to enable yaml parent/child binding on bus value. Error is generated if not matching or definition missing on parent side. Let me know if this answers the query
